### PR TITLE
Add Windows CI

### PR DESCRIPTION
**Pull request by joefitzgerald**
## Done
- [x] Add Windows CI via Appveyor
- [x] Pave the road for releases to be generated from Appveyor (Windows)
- [x] Make various improvements to script/cibuild and build/Gruntfile.coffee
- [x] Add caching of `~/.atom/.npm`, `~/.atom/.apm`, and `~/.atom/.node-gyp` on Appveyor
## Future
- [ ] Pave the road for releases to be generated from Travis (OSX / Linux)
- [ ] Provide a pathway for Node v4.2.1 / apm 1.5.x, but do not make them the default for internal Atom CI
- [ ] Add caching of `~/.atom/.npm`, `~/.atom/.apm`, and `~/.atom/.node-gyp` on Travis
- [ ] Make remoteReleases synchronization faster
- [ ] Add caching of electron
- [ ] Add caching of node in `https://github.com/atom/apm/blob/master/script/download-node.js`
- [ ] Create zip archive for Windows
- [ ] Add artifacts explicitly instead of all of `atom-build`
- [ ] Disable `0.10.35` builds on Windows (we now know they work)
- [ ] Add `signtool` tool
- [ ] Add encrypted ATOM_ACCESS_TOKEN to .travis.yml, with scopes necessary for release upload
- [ ] Fix .deb package generation
- [ ] Add encrypted signing keys for Travis and Appveyor, with corresponding encrypted variable

### CODE REVIEW

File: script/cibuild [hunk 1]
```diff
@@ -22,16 +22,18 @@ function loadEnvironmentVariables(filePath) {
...
-    loadEnvironmentVariables(path.resolve('/jenkins/config/atomcredentials'));
-  } else if (process.platform === 'darwin') {
-    loadEnvironmentVariables('/var/lib/jenkins/config/atomcredentials');
-    loadEnvironmentVariables('/var/lib/jenkins/config/xcodekeychain');
-  } else if (process.platform === 'linux') {
-    // Use Clang for building native code, the GCC on Precise is too old.
-    process.env.CC = 'clang';
-    process.env.CXX = 'clang++';
-    process.env.npm_config_clang = '1';
+  if (process.env.JANKY_SHA1) {
```
Comment (50Wliu) [inline script/cibuild]: :sparkles: Was thinking of creating a PR for this to get rid of the ENOENT errors on Travis.

**Comment by nathansobo** [comment]
Nice. Appreciate the task list for status.

File: .travis.yml [hunk 2]
```diff
@@ -14,23 +14,32 @@ compiler: clang
...
+  - npm config delete prefix
   - source /tmp/.nvm/nvm.sh
   - nvm install $NODE_VERSION
-  - nvm use $NODE_VERSION
+  - nvm use --delete-prefix $NODE_VERSION
+  - yes | cp apm/package-node-$NODE_VERSION.json apm/package.json
+  - yes | cp build/package-node-$NODE_VERSION.json build/package.json
 
-script: script/cibuild
+script: script/cibuild | ts
```
Comment (50Wliu) [inline .travis.yml]: `ts` doesn't seem to exist on Mac.  Is this even needed anyway?  All it does is add a timestamp which I don't think is particularly useful for CI.

**Comment by 50Wliu** [comment]
This would also probably be a good time to switch Travis's language to `node_js`.  Right now it's unset so it's defaulting to Ruby, which really isn't helping at all.  Switching to `node_js` should allow us to get rid of the manual nvm installation/Node configuration steps and let Travis handle it for us.

```yaml
language: node_js
node_js:
  - "4.2.1" # or alternatively, "stable"
  - "0.10.35" # 0.10 should work as well
```

**Comment by 50Wliu** [comment]
Appveyor also seems to support directly setting Node versions, though slightly differently from Travis: http://www.appveyor.com/docs/lang/nodejs-iojs

File: build/tasks/set-version-task.coffee [hunk 3]
```diff
@@ -5,10 +5,15 @@ module.exports = (grunt) ->
   {spawn} = require('./task-helpers')(grunt)
 
   getVersion = (callback) ->
-    onBuildMachine = process.env.JANKY_SHA1 and process.env.JANKY_BRANCH in ['stable', 'beta']
+    releasableBranches = ['stable', 'beta']
+    inCI = process.env.CI or process.env.JANKY_SHA1
```
Comment (50Wliu) [inline build/tasks/set-version-task.coffee]: Doesn't seem to be used

**Comment by joefitzgerald** [comment]
> Appveyor also seems to support directly setting Node versions, though slightly differently from Travis: http://www.appveyor.com/docs/lang/nodejs-iojs

Already using the technique described.

File: build/tasks/set-version-task.coffee (hunk 3 above)
Comment (joefitzgerald) [inline build/tasks/set-version-task.coffee]: Fixed in b4555a167a48bbea7ae6a5d0de07f9cf2a0edb82.

**Comment by joefitzgerald** [comment]
> This would also probably be a good time to switch Travis's language to node_js.

Done.

**Comment by joefitzgerald** [comment]
@50Wliu I have to back out use of the `language: node_js` support; it's not available on OSX yet: https://github.com/travis-ci/travis-ci/issues/2311.

**Comment by 50Wliu** [comment]
> it's not available on OSX yet: travis-ci/travis-ci#2311.

Ugh, whyyy

File: build/Gruntfile-ci.coffee [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+        options:
+          stdout: false
+          stderr: false
+          failOnError: false
+
+  grunt.registerTask('compile', ['babel', 'coffee', 'prebuild-less', 'cson', 'peg'])
+  grunt.registerTask('lint', ['standard', 'coffeelint', 'csslint', 'lesslint'])
+  grunt.registerTask('test', ['shell:kill-atom', 'run-specs'])
+
+  ciTasks = ['output-disk-space', 'download-electron', 'download-electron-chromedriver', 'build']
```
Comment (50Wliu) [inline build/Gruntfile-ci.coffee]: `output-disk-space` should be disabled for at least Travis, and maybe Appveyor.  It currently doesn't output anything on Travis.

Comment (joefitzgerald) [inline build/Gruntfile-ci.coffee]: Fixed in 6edb96ff9fc9af78b12356b357a914771fbd9790.

**Comment by nathansobo** [comment]
Kicked off a build on our internal CI infra so we can be sure we're still working there, too.

**Comment by nathansobo** [comment]
Kicked off another build after the force-push. Sorry, should have done this yesterday.

**Comment by nathansobo** [comment]
:zap: Thanks for your patience on this, both with the build and with us. Very excited about moving to more commodified build infra.

**Comment by maxbrunsfeld** [comment]
:clap: :moneybag: :chart_with_upwards_trend:

**Comment by mnquintana** [comment]
:rocket: :guitar: :clap:

**Pull request merged**